File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix-Solution1.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    return binarySearchMatrix(matrix, target, 0, (matrix.length * matrix[0].length) - 1);
};

const binarySearchMatrix = (matrix, target, left, right) => {
    if(left > right) return false;

    const middle = left + Math.floor((right - left) / 2);
    const row = Math.floor(middle / matrix[0].length);
    const col = middle % matrix[0].length;

    if(matrix[row][col] == target ) return true;
    if(matrix[row][col] < target) return binarySearchMatrix(matrix, target, left + 1, right);
    return binarySearchMatrix(matrix, target, left, right - 1);
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix-Solution1.cs
public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        return BinarySearchMatrix(matrix, target, 0, (matrix.Length * matrix[0].Length) - 1);
    }

    public bool BinarySearchMatrix(int[][] matrix, int target, int left, int right){
        if(left > right) return false;

        var middle = left + ((right - left) / 2);
        var row = middle / matrix[0].Length;
        var col = middle % matrix[0].Length;

        if(matrix[row][col] == target) return true;
        if(matrix[row][col] < target) return BinarySearchMatrix(matrix, target, left + 1, right);
        return BinarySearchMatrix(matrix, target, left, right - 1);
    }
}